Support reopening single files via unix domain socket in DaemonApp

nemanja-boric-sociomantic opened the pull request:
`DaemonApp`'s `ReopenableFilesExt` now supports reopening commands from the
UnixSocketExt. If you have the UnixSocketExt bound to the socket, sending command
`reopen_logfile file_path` will reopen this file, if registered, and acknowledge
the action with `ACK` or `ERROR` strings.

This is different than reopening files with the reopen signal, which
is not capable passing the file path to it, so it forces reopening all
files.

nemanja-boric-sociomantic committed:
Add File.path() method.

This adds path() method to the File class, used to get the File's
current path as cstring, without any GC activity.

nemanja-boric-sociomantic committed:
Support reopening single files via unix domain socket in DaemonApp

`DaemonApp`'s `ReopenableFilesExt` now supports reopening commands from the
UnixSocketExt. If you have the UnixSocketExt bound to the socket, sending command
`reopen_logfile file_path` will reopen this file, if registered, and acknowledge
the action with `ACK` or `ERROR` strings.

This is different than reopening files with the reopen signal, which
is not capable passing the file path to it, so it forces reopening all
files.

Fixes #190

nemanja-boric-sociomantic commented:
```
echo reopen_logfile /srv/dlsnode/log/root.log | sudo nc -U /srv/dlsnode/dlsnode.socket
ACK
```

gavin-norman-sociomantic commented on relnotes/reopensocket.feature.md:
```diff
@@ -0,0 +1,6 @@
+
+  `DaemonApp`'s `ReopenableFilesExt` now supports reopening commands from the
+  UnixSocketExt. If you have the UnixSocketExt bound to the socket, sending command
+  `reopen_logfile file_path` will reopen this file, if registered, and acknowledge
```
"will reopen the specified file"

nemanja-boric-sociomantic commented:
With this in place, reopening only affected files with the logrotate can be done as follows:

1. use `nosharedscripts` directive, so the post-rotate script is called once per file, with the absolute file path passed to it.
2. From the absolute file path, get the log directory and deduce socket placement from it (we could have convention that the socket is in the application's `run/` directory, so the socket path would be `
SOCKETPATH=$(dirname $LOGPATH)/../run/application.socket`
3. Simply execute:

```
test $(echo reopen_logfile $LOGPATH | sudo nc -U $SOCKETPATH) = ACK
```

and that should rotate your single logfile.

gavin-norman-sociomantic commented:
https://github.com/sociomantic-tsunami/ocean/pull/314#issuecomment-337876590 would be super useful in the release notes. We should try to make this feature as easy as possible for people to start using.

nemanja-boric-sociomantic commented:
Updated with the extended release notes.

gavin-norman-sociomantic commented on src/ocean/util/app/ext/ReopenableFilesExt.d:
```diff
@@ -70,14 +93,27 @@ public class ReopenableFilesExt : IApplicationExtension, ISignalExtExtension
     ***************************************************************************/
 
-    public this ( SignalExt signal_ext = null, int reopen_signal = SIGHUP )
+    public this ( SignalExt signal_ext = null,
```
Oh wait... this is a breaking change, or?

nemanja-boric-sociomantic replied:
Puh, yeah, I've missed that - I originally targeted v4.x.x. Just having new constructor with suffice, though. Will update.

leandro-lucarella-sociomantic replied:
Ah, I guess it's registered twice because of this... Maybe add an issue and a comment to clean this up in a following major?

gavin-norman-sociomantic reviewed (commented):
Submitting an early review because of the breaking change question.

nemanja-boric-sociomantic commented:
Updated with the dodged breaking change.

leandro-lucarella-sociomantic commented on src/ocean/io/device/File.d:
```diff
@@ -412,6 +412,18 @@ class File : Device, Device.Seek, Device.Truncate
+        public cstring path ()
+        {
+            return this.path_;
+        }
```
Why isn't this done in `toString()` too?

nemanja-boric-sociomantic replied:
Because `toString()` must return `istring` (from `Object`).

leandro-lucarella-sociomantic commented on relnotes/reopensocket.feature.md:
```diff
@@ -0,0 +1,20 @@
+  `DaemonApp`'s `ReopenableFilesExt` now supports reopening commands from the
+  UnixSocketExt. If you have the UnixSocketExt bound to the socket, sending
+  command `reopen_logfile file_path` will reopen the specified file, if
+  registered, and acknowledge the action with `ACK` or `ERROR` strings.
```
I wonder if, for debugging purposes, a message should be returned with the `ERROR` too (like `ERROR <free text>`).

nemanja-boric-sociomantic replied:
I could return ERROR + the error text, no problem.

leandro-lucarella-sociomantic commented on relnotes/reopensocket.feature.md:
```diff
@@ -0,0 +1,20 @@
+       directory, so the socket path would be
+       `SOCKETPATH=$(dirname $LOGPATH)/../application.socket`.
+    3) Finally, in post-rotate script, simply executing
+       `test $(echo reopen_logfile $LOGPATH | nc -q 2 -U $SOCKETPATH) = ACK` will
```
You probably should add quotes to `$(echo ...)`, otherwise it could be an empty string and make `test` give an error, right?

nemanja-boric-sociomantic replied:
Yeah.

nemanja-boric-sociomantic replied:
Good catch!

leandro-lucarella-sociomantic commented on src/ocean/util/app/DaemonApp.d:
```diff
@@ -380,6 +375,12 @@ public abstract class DaemonApp : Application,
+            this.unix_socket_ext,
+            settings.reopen_signal);
+        this.registerExtension(this.reopenable_files_ext);
+
```
Why is this now registered twice?

nemanja-boric-sociomantic replied:
Probably accident - we need to register it only once. Will fix.

nemanja-boric-sociomantic replied:
Wait, where do you see it's registered twice?

nemanja-boric-sociomantic replied:
I've just moved it to be registered after UnixSocketExt, since we need it for registering the handler.

leandro-lucarella-sociomantic replied:
I guess I was colorblind and didn't see the red above... All good.

leandro-lucarella-sociomantic commented on src/ocean/util/app/ext/ReopenableFilesExt.d:
```diff
@@ -70,14 +94,54 @@ public class ReopenableFilesExt : IApplicationExtension, ISignalExtExtension
+    public this ( SignalExt signal_ext,
+        UnixSocketExt unix_socket_ext,
+        int reopen_signal = SIGHUP,
+        cstring reopen_command = "reopen_logfile")
```
Do we really want to make this configurable? It seems strange to be able to introduce inconsistencies in the socket interface...

nemanja-boric-sociomantic replied:
Well, we do the same thing for the SIGHUP. Note that people don't actually use this class directly, but it's used withing DaemonApp.

mathias-lang-sociomantic replied:
Travis says this should be an `istring`
Also this works with *any* file opened with `ReusableFilesExt` right ? In this case `logfile` might be a bit too restrictive of a name.

mathias-lang-sociomantic replied:
Another comment about this: It looks weird to have signal-related thing, unixsocket, signal-related, unix-socket.

I understand why you went with this, and the alternative is not too appealing as well, so I'm going to propose an alternate solution: since both `SignalExt` and `UnixSocketExt` are optional and independent of each other, why don't we move the logic to `setup` methods, e.g. `setupUnixSocket(UnixSocketExt, istring = "reopen_file")` and `setupSignalExt(SignalExt, int = SIGHUP)` (naming up for discussion).

nemanja-boric-sociomantic replied:
Right, deprecating constructor and having setup methods replace it would work, good idea :+1:

nemanja-boric-sociomantic replied:
Damn, I can't deprecate the existing constructor as all the methods are optional, so simply `auto ext = new ReopenableFilesExt` will trigger the deprecation :(.

leandro-lucarella-sociomantic commented on src/ocean/util/app/ext/ReopenableFilesExt.d:
```diff
@@ -130,6 +224,27 @@ public class ReopenableFilesExt : IApplicationExtension, ISignalExtExtension
+    private void socketReloadCommand ( cstring[] args,
+            void delegate ( cstring response ) send_response )
+    {
+        // Todo: reopen the file here.
```
I guess this `TODO` is outdated?

nemanja-boric-sociomantic replied:
Oh, right.

nemanja-boric-sociomantic commented:
Updated.

leandro-lucarella-sociomantic reviewed (approved):
All the pending stuff is minor, so approving, feel free to merge and then work on any improvements.

mathias-lang-sociomantic commented:
Any chance we could get this tested by the right people ( :wave: @andrew-stevenson-sociomantic ) so we are sure it fulfills their use-case ? :)
Thanks a lot for taking care of this @nemanja-boric-sociomantic 
Fixes https://github.com/sociomantic-tsunami/ocean/issues/190

mathias-lang-sociomantic commented on src/ocean/util/app/ext/ReopenableFilesExt.d:
```diff
@@ -130,6 +224,37 @@ public class ReopenableFilesExt : IApplicationExtension, ISignalExtExtension
+            return;
+        }
+
+        auto filename = args[0];
```
Could we use a `while` loop here, so one can pass multiple files ?

nemanja-boric-sociomantic replied:
I didn't anticipated that one can do that, since logrotate doesn't do it, but it makes sense, I'll update it.

mathias-lang-sociomantic commented on src/ocean/util/app/ext/ReopenableFilesExt.d:
```diff
@@ -70,14 +94,54 @@ public class ReopenableFilesExt : IApplicationExtension, ISignalExtExtension
+        this.cwd = Environment.directory();
+
+        // Environment.directory makes sure for the trailing /, confirm it
+        verify (this.cwd[$-1] == '/');
```
Extra space

mathias-lang-sociomantic replied:
Also the comment sounds weird to me

nemanja-boric-sociomantic replied:
Yes, should be `Environment.directory should always return trailing /, confirm that that's the case`

mathias-lang-sociomantic commented on relnotes/reopensocket.feature.md:
```diff
@@ -0,0 +1,20 @@
+    3) Finally, in post-rotate script, simply executing
+       `test "$(echo reopen_logfile $LOGPATH | nc -U $SOCKETPATH)" = ACK` will
+       instruct the application to reopen the given logfile and report status
+       to logrotate.
```
Maybe give a very simple sample file ?

mathias-lang-sociomantic commented on src/ocean/util/app/ext/ReopenableFilesExt.d:
```diff
@@ -70,14 +94,54 @@ public class ReopenableFilesExt : IApplicationExtension, ISignalExtExtension
 
-    public this ( SignalExt signal_ext = null, int reopen_signal = SIGHUP )
+    public this ( SignalExt signal_ext = null,
+        int reopen_signal = SIGHUP )
```
Why the change indent ?

mathias-lang-sociomantic replied:
Since this is internal-only, should it be deprecated ?

nemanja-boric-sociomantic replied:
Interesting thing that this is not internal only - once could create their completely separate application hierarchy and use this. However, I will deprecate this, good catch :+1:

Indentation was changed via copy/pasting from other constructor. Will fix.

leandro-lucarella-sociomantic commented:
BTW, adding some test would be nice. It is a bit annoying but it should be possible doing something like: create extension, register file X, write, move file externally to X and create a new file X, write again, check it's written to Y (and not to X), reopen files, write again and check it's written to X (and not to Y).

nemanja-boric-sociomantic commented:
Yeah, the test could/should be added.

mihails-strasuns-sociomantic commented on src/ocean/util/app/ext/ReopenableFilesExt.d:
```diff
@@ -114,6 +178,36 @@ public class ReopenableFilesExt : IApplicationExtension, ISignalExtExtension
+    {
+        foreach (file; this.open_files)
+        {
+            path_buffer.length = 0;
```
I'd recommend `path_buffer.reset()`, it makes it more obvious you don't need `enableStomping` :)

mathias-lang-sociomantic replied:
Yeah I started typing `enableStomping` and had the "oh..." moment :P

nemanja-boric-sociomantic replied:
The sole purpose of Buffer :-P

nemanja-boric-sociomantic commented:
Updated with all points except the example logrotate file and test.

nemanja-boric-sociomantic commented:
Updated with the test.

nemanja-boric-sociomantic commented:
Updated with the example logrotate config file and the fact that there may be multiple files registered with the same filename, and they all needs to be reloaded.

nemanja-boric-sociomantic commented:
I think this includes all requested changes on this PR.

nemanja-boric-sociomantic committed:
Don't allocate on reopening of the logfiles

File.toString must return immutable string, so this
always allocates. Instead, using File.path() which returns
cstring view to the path is enough.

nemanja-boric-sociomantic commented:
Updated with the last bit: `ReopenableFilesExt` was getting the file's path via `File.toString` which would always allocate, given it needs to return `istring`. It's now tweaked to use newly added `File.path`.

mathias-lang-sociomantic reviewed (approved):
LGTM, pending testing I assume ?

nemanja-boric-sociomantic commented:
Yes, I've tested it with the DLS node and it works. I wonder if @gautam-kotian-sociomantic could help us test this as well, if @andrew-stevenson-sociomantic doesn't have much time.

If nobody has time for a test, I think it's fair to merge this as is, given that I did a bit of testing, and fix issues in the patch releases, than to leave this PR rotting.

andrew-stevenson-sociomantic commented:
> if @andrew-stevenson-sociomantic doesn't have much time.

I suspect it will take a lot of time for me to test as I would need to update ocean, add command socket support to an app, switch from our logging stuff to ocean and then test :-(

nemanja-boric-sociomantic commented:
Just a note there: adding command socket support is done automatically with the recent oceans (just add `[UNIX_SOCKET]path=app.path` in the config file). But yeah, removing the custom stuff is more problematic.

leandro-lucarella-sociomantic commented:
If we have unit/integration tests and simple test seems to work, I would just merge.

nemanja-boric-sociomantic commented:
Yes, the intergration test has been added.

leandro-lucarella-sociomantic commented:
I don't get why the coverage is so crappy now. Lately I couldn't make sense of the coverage reports :angry: 

Also, needs rebase now.

nemanja-boric-sociomantic commented:
The biggest problem we have IMHO is that the coverage is reported file per file only. So, having something tested in the different file (like we're doing in the integration tests), doesn't help coverage at all :(. The second, separate problem, which is way more annoying is that these unexpected coverage changes. I wonder if that's related to the fact that we started uploading coverage for all matrix items (and actually building stuff with F=production), so codecov is going crazy because of that.

nemanja-boric-sociomantic commented:
Rebased.

leandro-lucarella-sociomantic commented:
> I wonder if that's related to the fact that we started uploading coverage for all matrix items (and actually building stuff with F=production), so codecov is going crazy because of that.

Mmm, maybe it is worth trying to disable codecov for F=production builds then. It's really annoying right now.
